torch/ao/pruning/sparsifier/utils.py

```diff
@@ -1,16 +1,70 @@
 from typing import Any, Dict, Optional
+from torch.nn.utils.parametrize import type_before_parametrizations, is_parametrized
+from itertools import chain
 
 from torch import nn
 
 __all__ = [
+    "module_contains_param",
+    "swap_module",
     "module_to_fqn",
     "fqn_to_module",
     "get_arg_info_from_tensor_fqn",
     "FakeSparsity",
 ]
 
 
-def module_to_fqn(model: nn.Module, module: nn.Module, prefix: str = "") -> Optional[str]:
+def module_contains_param(module, parametrization):
+    if is_parametrized(module):
+        # see if any of the module tensors have a parametriztion attached that matches the one passed in
+        return any(
+            [
+                any(isinstance(param, parametrization) for param in param_list)
+                for key, param_list in module.parametrizations.items()
+            ]
+        )
+    return False
+
+
+def swap_module(mod, mapping):
+    r"""Swaps the module if it has a quantized counterpart and it has an
+    `observer` attached.
+    Args:
+        mod: input module
+        mapping: a dictionary that maps from nn module to sparse nn module
+    Return:
+        The corresponding sparse module of `mod` according to mapping
+    """
+    new_mod = mod
+    if type_before_parametrizations(mod) in mapping:
+        sparse_mod = mapping[type_before_parametrizations(mod)]
+        new_mod = sparse_mod.from_dense(mod)
+
+        # Preserve module's pre forward hooks. They'll be called on quantized input
+        for pre_hook_fn in mod._forward_pre_hooks.values():
+            new_mod.register_forward_pre_hook(pre_hook_fn)
+        # Preserve module's post forward hooks except _observer_forward_hook
+        # After convert they'll work with quantized output
+        for hook_fn in mod._forward_hooks.values():
+            if hook_fn is not _observer_forward_hook:
+                new_mod.register_forward_hook(hook_fn)
+
+        # respect device affinity when swapping modules
+        devices = {p.device for p in chain(mod.parameters(), mod.buffers())}
+        assert len(devices) <= 1, (
+            "swap_module only works with cpu or single-device CUDA modules, "
+            "but got devices {}".format(devices)
+        )
+        device = next(iter(devices)) if len(devices) > 0 else None
+        if device:
+            new_mod.to(device)
```

Comment:
nit: looks like this is copied from eager mode quantization? maybe you can refactor this to a separate helper function?

Reply:
I'll add this to the BE tracker and take another look later. 

I'm not sure atm what the best way to reuse the shared code other than passing a function around for the sparse/quant specific steps. And I don't like that b/c I feel like it makes the code harder to read.

Not sure if you have any opinions on this?